File: 01.factory-method/factoryMethod.js
// Producto
class Button {
    render() {
        throw new Error('Método render() debe ser implementado.');
    }
}

// Producto Concreto
class WindowsButton extends Button {
    render() {
        console.log('Renderizando botón de Windows.');
    }
}

// Producto Concreto
class MacOSButton extends Button {
    render() {
        console.log('Renderizando botón de MacOS.');
    }
}

// Creador
class Dialog {
    renderWindow() {
        const okButton = this.createButton();
        okButton.render();
    }
    createButton() {
        throw new Error('Método createButton() debe ser implementado.');
    }
}

// Creador Concreto
class WindowsDialog extends Dialog {
    createButton() {
        return new WindowsButton();
    }
}

// Creador Concreto
class MacOSDialog extends Dialog {
    createButton() {
        return new MacOSButton();
    }
}

// Uso
const os = require('os').platform();
let dialog;
if (os === 'win32') {
    dialog = new WindowsDialog();
} else {
    dialog = new MacOSDialog();
}
dialog.renderWindow();

File: 01.factory-method/factoryMethod.java
// Producto
interface Button {
    void render();
}

// Producto Concreto
class WindowsButton implements Button {
    public void render() {
        System.out.println("Renderizando botón de Windows.");
    }
}

// Producto Concreto
class MacOSButton implements Button {
    public void render() {
        System.out.println("Renderizando botón de MacOS.");
    }
}

// Creador
abstract class Dialog {
    public void renderWindow() {
        Button okButton = createButton();
        okButton.render();
    }
    protected abstract Button createButton();
}

// Creador Concreto
class WindowsDialog extends Dialog {
    protected Button createButton() {
        return new WindowsButton();
    }
}

// Creador Concreto
class MacOSDialog extends Dialog {
    protected Button createButton() {
        return new MacOSButton();
    }
}

// Uso
public class factoryMethod {
    public static void main(String[] args) {
        Dialog dialog;
        String os = System.getProperty("os.name").toLowerCase();
        if (os.contains("win")) {
            dialog = new WindowsDialog();
        } else {
            dialog = new MacOSDialog();
        }
        dialog.renderWindow();
    }
}